package/MDAnalysis/analysis/Aggregation_size.py

```diff
@@ -0,0 +1,370 @@
+        """
+        This is a subfunction that enables the code to find all
+        connected molecules in a cluster.
+        k is a connected molecule to the main molecule in the first level.
+        self._frame_index is a time frame.
+        counter is a step.
+        m provides a suitable off_set to match the number of molecules.
+        Limit is the cut_off provided by the user.
+        """
+        bool_sector = self.d[k] < cut_off
+        Coordinates = self.d[k][bool_sector]
+        # This "if" will check if there is any uninvestigated molecule left
+        if len(Coordinates) != 0:
+            index1 = np.where(bool_sector)
+            listOfCoordinates = list(zip(index1[0], index1[1]))
+            connect = ([j[0] for j in listOfCoordinates]
+                       + connected[m] + 1)
+            Connected_NLevel = np.array(connect)
+        else:
+            Connected_NLevel = []
+        return Connected_NLevel
+
+    def _single_frame(self):
+        """
+        1- Import the needed module
+        2- Calculate the distance between molecules based on the intended
+        criteria
+        3- Find the aggregates by checking only one time each using the
+        "aggregate" subfunction (to be efficient)
+        4- Check and merge two aggregates as one aggregate if the distance
+        between any element of them is less than cut_off
+        5- Eliminate monomers (optional)
+        6- measure average gyration radius (optional)
+        7- measure the number of aggregates with different size
+        """
+        aggregate = [0] * self._molnum
+        if self._type == "Closest":
+            self.closest_atom(self.d)
+        elif self._type == "COM":
+            self.center_of_mass(self.d)
+        elif self._type == "Specific":
+            self.specific_atoms(self.d)
+        else:
+            raise ValueError("enter valid Type")
+        # P_Level collact all the investigated nodes in past levels
+        P_Level = []
+        # Group will collect the molecule first investigating molecule
+        # len(Group) will be the number of clusters
+        Group = []
+        # this for will go through all molecules to
+        # check their connection with others
+        for counter in range(1, self._molnum):
+            # aggergate parameter include the molecules in each cluster
+            aggregate[counter-1] = [counter]
+            # this "if" will restrict double calculation of the same molecule
+            # to increase efficincy of code
+            if counter not in P_Level:
+                # define the ones that are attached
+                # to a molecule in first level
+                bool_sector = self.d[counter] < self._cut_off
+                Coordinates = self.d[counter][bool_sector]
+                index1 = np.where(bool_sector)
+                # list the connected molecules to a molecule in first level
+                listOfCoordinates = list(zip(index1[0], index1[1]))
```

Comment:
When I see `list` and `tolist()` I wonder if this code can be written with arrays instead. Big lists are typically slow and the wrong data structure for coordinates.

Reply:
Also, `listofCoordinates` is again confusing, as it actually contains the indices of the residues with a sufficiently low cutoff. Although I think this will error as there should not be a second `index[1]` element in a 1D array.

Reply:
Sorry, I guess the bad name for parameter cause some misunderstanding. Actually, I do not have any molecule coordinate after measuring distances. I have changed the names and make them more rational. So here we only have distances between molecules. I agree with you that the arrays are faster and more suitable for coordination and distance calculation, that is why I used them at the beginning of the code. I agree that the code will be faster if everything will be with arrays but the point is some times I used dictionaries and sets which are more compatible with list. So if you do not mind, I will keep the code as it is in terms of this comment and make test. If after the test section you still think code is not efficient I will try my best to rewrite the lists with arrays.